napari/utils/key_bindings.py

```diff
@@ -182,6 +186,284 @@ def normalize_key_combo(key_combo):
 UNDEFINED = object()
 
 
+def call_with_context(function, context):
+    """
+    call function `function` with the corresponding value taken from context
+
+    This is use in the action manager to pass the rights instances to the actions,
+    without having the need for them to take a **kwarg, and is mostly needed when
+    triggering actions via buttons, or to record.
+
+    If we went a declarative way to trigger action we cannot refer to instances
+    or objects that must be passed to the action, or at least this is
+    problematic.
+
+    We circumvent this by having a context (dictionary of str:instance) in
+    the action manager, and anything can tell the action manager "this is the
+    current instance a key". When an action is triggered; we inspect the
+    signature look at which instances it may need and pass this as parameters.
+    """
+    from inspect import signature
+
+    context_keys = [
+        k
+        for k, v in signature(function).parameters.items()
+        if v.kind not in (v.VAR_POSITIONAL, v.VAR_KEYWORD)
+    ]
+    ctx = {k: v for k, v in context.items() if k in context_keys}
+    return function(**ctx)
+
+
+@dataclass
+class Action:
+    command: Callable
+    description: str
+    keymappable: Any
+
+
+class ActionManager:
+    """
+    Manage the bindings between buttons; shortcuts, callbacks gui elements...
+
+    The action manager is aware of the various buttons, keybindings and other
+    elements that may trigger an action and is able to synchronise all of those.
+    Thus when a shortcut is bound; this should be capable of updating the
+    buttons tooltip menus etc to show the shortcuts, descriptions...
+
+    In most cases this should also allow to bind non existing shortcuts,
+    actions, buttons, in which case they will be bound only once the actions are
+    registered.
+
+    As the action manager also knows about all the UI elements that can trigger
+    an action, it can also be used to flash the buttons when the action is
+    triggered separately.
+
+    For actions that need access to a global element (a viewer, a plugin, ... ),
+    you want to give this item a unique name, and add it to the action manager
+    `context` object.
+
+    >>> action_manager.context['number'] = 1
```

Comment:
The potential failure modes here with the `.context` attribute are still a little unclear to me.  do we need to have any sort of convention to the context keys?  Is it something that only napari is going to use?  or end-users also?

Reply:
I think that this will mostly be napari internal object, like `viewer`, `active_layer(s)` plugins may expose some if they have instances on objects actions that needs instances of some objects that may change.